lib/pam/pam.go

```diff
@@ -0,0 +1,326 @@
+}
+
+var handlerMu sync.Mutex
+var handlerCount int
+var handlers map[int]handler = make(map[int]handler)
+
+//export writeCallback
+func writeCallback(index C.int, stream C.int, s *C.char) {
+	handlerMu.Lock()
+	defer handlerMu.Unlock()
+
+	handle, ok := handlers[int(index)]
+	if !ok {
+		log.Errorf("Unable to write to output stream, no handler for index %v found", int(index))
+		return
+	}
+
+	// To prevent poorly written PAM modules from sending more data than they
+	// should, cap strings to the maximum message size that PAM allows.
+	str := C.GoStringN(s, C.int(C.strnlen(s, C.PAM_MAX_MSG_SIZE)))
+
+	// Write to the stream (typically stdout or stderr or equivlient).
+	handle.writeStream(int(stream), str)
+}
+
+//export readCallback
+func readCallback(index C.int, e C.int) *C.char {
+	handlerMu.Lock()
+	defer handlerMu.Unlock()
+
+	handle, ok := handlers[int(index)]
+	if !ok {
+		log.Errorf("Unable to read from input stream, no handler for index %v found", int(index))
+		return nil
+	}
+
+	var echo bool
+	if e == 1 {
+		echo = true
+	}
+
+	// Read from the stream (typically stdin or equivlient).
+	s, err := handle.readStream(echo)
+	if err != nil {
+		log.Errorf("Unable to read from input stream: %v", err)
+		return nil
+	}
+
+	// Return one less than PAM_MAX_RESP_SIZE to prevent a Teleport user from
+	// sending more than a PAM module can handle and to allow space for \0.
+	//
+	// Note: The function C.CString allocates memory using malloc. The memory is
+	// not released in Go code because the caller of the callback function (PAM
+	// module) will release it. C.CString will null terminate s.
+	n := int(C.PAM_MAX_RESP_SIZE)
+	if len(s) > n-1 {
+		return C.CString(s[:n-1])
+	}
+	return C.CString(s)
+}
+
+// registerHandler will register a instance of *PAM with the package level
+// handlers to support callbacks from C.
+func registerHandler(p *PAM) int {
+	handlerMu.Lock()
+	defer handlerMu.Unlock()
+
+	// The make_pam_conv function allocates struct pam_conv on the heap. It will
+	// be released by Close function.
+	handlerCount = handlerCount + 1
+	p.conv = C.make_pam_conv(C.int(handlerCount))
+	handlers[handlerCount] = p
+
+	return handlerCount
+}
+
+// unregisterHandler will remove the PAM handle from the package level map
+// once no more C callbacks can come back.
+func unregisterHandler(handlerIndex int) {
+	handlerMu.Lock()
+	defer handlerMu.Unlock()
+
+	delete(handlers, handlerIndex)
+}
+
+var buildHasPAM bool = true
+var systemHasPAM bool = false
+
+// pamHandle is a opaque handle to the libpam object.
+var pamHandle unsafe.Pointer
+
+func init() {
+	// Obtain a handle to the PAM library at runtime. The package level variable
+	// SystemHasPAM is updated to true if a handle is obtained.
+	//
+	// Note: Since this handle is needed the entire time Teleport runs, dlclose()
+	// is never called. The OS will cleanup when the process exits.
+	pamHandle = C.dlopen(C.library_name(), C.RTLD_NOW)
+	if pamHandle != nil {
+		systemHasPAM = true
```

Comment:
there is a potential race here, you have to guard the systemHasPAM reads and writes with mutexes or use atomic